### ISSUE TITLE
support multiple contract source path

### ISSUE THREAD

**Issue by wighawag** [NomicFoundation/hardhat#520]
Buidler sources setting for contract source path is currently just a string argument. It does not support having contract code spread into multiple folder.

Sometime, it is beneficial to break a project into separate git module. While it is possible to have the submodule as a sub folder in the project src path, it has the issue that every file of that submodule being present in the source path.

If buidler could support an array of path for contracts source path, this would aloow a project to have  other contract source as submodule in different folder.

**Comment by tuler** [comment]
Consider the following scenario:

Project A
Run `npx builder node`.
Project A contracts get automatically deployed.

Project B
Run `npx builder deploy --network localhost`
Project B contracts are deployed, but the EVM prints:

```
eth_sendTransaction
  Contract deployment: <UnrecognizedContract>
  Contract address:    0x500d1d6a4c7d8ae28240b47c8fcde034d827fd5e
  Transaction:         0xf38257dd487f5eba4e3af9a9a30b931aab982e5ae944de54bd331daa8af87249
  From:                0xc783df8a850f42e7f7e57013759c285caa701eb6
  Value:               0 ETH
  Gas used:            3956533 of 9024999
  Block #9:            0x79f93510eaddbc8eb5bedfb0426ee28b90e8e16ccb49e2fbd6c3c420fe407149
```

Is this the cause of `<UnrecognizedContract>`?

**Comment by fvictorio** [comment]
@tuler I don't think this issue could cause your problem. If you can share how to reproduce it with the latest buidler version, please open a new issue.

**Comment by cameel** [comment]
Just a head up - a similar feature will be soon added to the compiler for CLI use: https://github.com/ethereum/solidity/issues/11409. It won't affect Hardhat since Hardhat uses Standard JSON input and finds import files on its own but if it's ever implemented here it would be nice to have these features somewhat consistent with each other. Especially regarding the conversion from absolute to relative paths.

**Comment by fubhy** [comment]
Agreed that this would be nice. For instance, our core protocol architecture comes with three different extension systems. Each of these systems can have an arbitrary number of plugins (adapters into other protocols, etc.) and it would be nice if we could split them up into seperate packages within a monorepo architecture. I'm happy to work on this if that's something you'd like to support.

Instead of allowing the `paths` to be specified as both an array of strings or a plain string I'd argue that it would be better to allow users to define multiple relative project roots similar to how e.g. Jest or other tools enable you to configure multiple projects in one configuration file.

**Comment by samlaf** [comment]
@cameel are you suggesting that one would run `npx hardhat compile` with a `--include-paths` flag?
What's the status on such a feature? Hardhat cli has very few features at the moment:
```
Hardhat version 2.9.7

Usage: hardhat [GLOBAL OPTIONS] compile [--concurrency <INT>] [--force] [--no-typechain] [--quiet]

OPTIONS:

  --concurrency         Number of compilation jobs executed in parallel. Defaults to the number of CPU cores - 1 (default: 7)
  --force               Force compilation ignoring cache 
  --no-typechain        Skip Typechain compilation 
  --quiet               Makes the compilation process less verbose 

compile: Compiles the entire project, building all artifacts

For global options help run: hardhat help
```

**Comment by samlaf** [comment]
> Hardhat has its own independent mechanism for finding Solidity files outside of your main project dir. What I wrote was just a bit of tangential info that might be interesting to whoever will be implementing it.

Are you saying that this issue was resolved? Because the ts files I was looking at were still only accepting a single string as argument for `sources` and `artifacts` folders (even if the variable names have an s at the end).

**Comment by cameel** [comment]
> Are you saying that this issue was resolved?

No, I didn't say that. The issue is still open.

**Comment by fvictorio** [comment]
Closing in favor of #776.
